Provide partial support for string "format" constraint

stefanov-sm opened the pull request:
### Provide partial support for string ["format"](https://json-schema.org/understanding-json-schema/reference/string#format) constraint

Formats that correspond to native PostgreSQL data types are implemented. These are

> date-time, date, time, duration, uuid, ipv4, ipv6, regex  

Email format validation by a regex suggested by [pbaumard](https://github.com/pbaumard).

Consistent with current behaviour unsupported options validate positive.

stefanov-sm committed:
Implement partial support of "format" for type "string"

stefanov-sm committed:
Provide partial support for string "format" constraint

stefanov-sm committed:
Provide partial support for string "format" constraint

stefanov-sm commented:
"email" format validation may be added with one extra line in the `case` list that uses the popular regular expression:
```sql
 WHEN 'email' THEN IF target !~* '^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$' THEN RAISE; END IF;
```
It would not be as strict as the rest though. The regular expression might be brushed up too.

kwakwaversal commented:
Nice work. I've thought about adding this myself and can see it might be useful (especially for uuid).

Have you performance tested this though? Something in the back of my head thinks that adding exception handling adds a bit of overhead.

stefanov-sm commented:
Fair enough, exception handling does add some overhead. Yet given the relative complexity of regular expressions and ISO8601 validation (620+ timezone names and abbreviations) I think that this is a price worth paying. The performance loss is not that significant. Benchmark on a very modest laptop - 100K `{"type":"number"}` validate in 11.4 s, 100K `{"type":"string", "format":"date-time"}` validate in 12.9 s.

kwakwaversal commented:
Thanks for checking. What are the with exception handling and without exception handling benchmark times?

It's worth mentioning because people might validate large amounts of data (if triggers run on updates for example) which might have unintended consequences based on the current version.

pbaumard commented:
> "email" format validation may be added with one extra line in the `case` list that uses the popular regular expression:
> 
> ```sql
>  WHEN 'email' THEN IF target !~* '^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$' THEN RAISE; END IF;
> ```
> 
> It would not be as strict as the rest though. The regular expression might be brushed up too.

Using the regexp in https://dba.stackexchange.com/a/165923 following [HTML5 email spec](https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type=email)) would surely be better:
```sql
WHEN 'email' THEN IF target !~ '^[a-zA-Z0-9.!#$%&''*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$' THEN RAISE; END IF;
```
it does not follow [RFC 5321, section 4.1.2](http://tools.ietf.org/html/rfc5321#section-4.1.2) specified in [json-schema format](https://json-schema.org/understanding-json-schema/reference/string#format), but as written in [HTML5 email spec](https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type=email)):

> This requirement is a [willful violation](https://html.spec.whatwg.org/multipage/introduction.html#willful-violation) of RFC 5322, which defines a syntax for email addresses that is simultaneously too strict (before the "@" character), too vague (after the "@" character), and too lax (allowing comments, whitespace characters, and quoted strings in manners unfamiliar to most users) to be of practical use here.

stefanov-sm commented:
Validating 100K `{"type":"number"}` with commented `"format"` section (i.e. the original function with no exception handling code) took the same 11.3 s. There is no performance change unless `{"type":"string", "format":"date-time"}` is hit and the exception machinery gets invoked.

stefanov-sm committed:
"email" format added, regex implementation 

Uses the HTML5-style regex suggested by [Pierre Baumard](https://github.com/pbaumard)

stefanov-sm commented:
> > "email" format validation may be added with one extra line in the `case` list that uses the popular regular expression:
> > ```sql
> >  WHEN 'email' THEN IF target !~* '^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$' THEN RAISE; END IF;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It would not be as strict as the rest though. The regular expression might be brushed up too.
> 
> Using the regexp in https://dba.stackexchange.com/a/165923 following [HTML5 email spec](https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type=email)) would surely be better:
> 
> ```sql
> WHEN 'email' THEN IF target !~ '^[a-zA-Z0-9.!#$%&''*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$' THEN RAISE; END IF;
> ```
> 
> it does not follow [RFC 5321, section 4.1.2](http://tools.ietf.org/html/rfc5321#section-4.1.2) specified in [json-schema format](https://json-schema.org/understanding-json-schema/reference/string#format), but as written in [HTML5 email spec](https://html.spec.whatwg.org/multipage/input.html#e-mail-state-(type=email)):
> 
> > This requirement is a [willful violation](https://html.spec.whatwg.org/multipage/introduction.html#willful-violation) of RFC 5322, which defines a syntax for email addresses that is simultaneously too strict (before the "@" character), too vague (after the "@" character), and too lax (allowing comments, whitespace characters, and quoted strings in manners unfamiliar to most users) to be of practical use here.

Nice & clean. I have added it to the PR, quoting your comment/suggestion.

pbaumard commented on postgres-json-schema--0.1.1.sql:
```diff
@@ -237,6 +237,28 @@ BEGIN
+  IF schema ? 'format' AND jsonb_typeof(data) = 'string' THEN
+    DECLARE
+      target text := (data #>> '{}');
+      EMAIL_RX constant text := '^[\w.!#$%&''*+/=?^`{|}~-]+@[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?(?:\.[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?)*$';
```
Case sensitive check is faster, so why not use the full case sensitive regex here and `target !~ EMAIL_RX` below?

```suggestion
      EMAIL_RX constant text := '^[a-zA-Z0-9.!#$%&''*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$';
```

stefanov-sm replied:
Fine. Longer and maybe more difficult to read but 10% are always 10%.

pbaumard commented on postgres-json-schema--0.1.1.sql:
```diff
@@ -237,6 +237,28 @@ BEGIN
+        WHEN 'ipv6'      THEN PERFORM target::inet; IF target NOT LIKE '%:%' THEN RAISE; END IF;
+        WHEN 'ipv4'      THEN PERFORM target::inet; IF target LIKE '%:%' THEN RAISE; END IF;
+        WHEN 'regex'     THEN PERFORM '' ~ target; 
+        WHEN 'email'     THEN IF target !~* EMAIL_RX THEN RAISE; END IF;
```
See regex comment above, a case sensitive check would be faster:
```suggestion
        WHEN 'email'     THEN IF target !~ EMAIL_RX THEN RAISE; END IF;
```

stefanov-sm committed:
Case-sensitive mail gegex suggested by [pbaumard](https://github.com/pbaumard)
